xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -58,521 +66,888 @@ define(["jquery", "polyfills"], function ($) {
+     * @param {String} propertyId
+     */
+    isValidPropertyId: function (propertyId) {
+      return this.getPropertyIds().indexOf(propertyId) !== -1;
+    },
+
+
+
+
+
+    /**
+     * ---------------------------------------------------------------
+     * DESCRIPTORS
+     */
+
+
+    /**
+     * Return the property descriptor corresponding to a property id
+     * @param {String} propertyId
+     * @returns {Object}
+     */
+    getPropertyDescriptor: function (propertyId) {
+      return this.data.meta.propertyDescriptors.find(function (propertyDescriptor) {
+        return propertyDescriptor.id === propertyId;
+      });
+    },
+
+
+    /**
+     * Return the property type descriptor corresponding to a property id
+     * @param {String} propertyId
+     * @returns {Object}
+     */
+    getPropertyTypeDescriptor: function (propertyId) {
+      var propertyDescriptor = this.getPropertyDescriptor(propertyId);
+      if (!propertyDescriptor) { return; }
+      return this.data.meta.propertyTypes.find(function (typeDescriptor) {
+        return typeDescriptor.id === propertyDescriptor.type;
+      });
+    },
+
+
+    /**
+     * Return the property type descriptor corresponding to a property id
+     * @param {String} propertyId
+     * @returns {Object}
+     */
+    getLayoutDescriptor: function (layoutId) {
+      return this.data.meta.layoutDescriptors.find(function (layoutDescriptor) {
+        return layoutDescriptor.id === layoutId;
+      });
+    },
+
+
+    /**
+     * Get the displayer descriptor associated to a property id
+     * @param {String} propertyId
+     * @returns {Object}
+     */
+    getDisplayerDescriptor: function (propertyId) {
+      if (!this.isValidPropertyId(propertyId)) { return; }
+      // property descriptor config
+      var propertyDescriptor = this.getPropertyDescriptor(propertyId) || {};
+      var propertyDescriptorDisplayer = propertyDescriptor.displayer || {};
+      // property type descriptor config
+      var typeDescriptor = this.getPropertyTypeDescriptor(propertyId) || {};
+      var typeDescriptorDisplayer = typeDescriptor.displayer || {};
+      // merge property and/or type displayer descriptors
+      var highLevelDisplayer;
+      if (!propertyDescriptorDisplayer.id || propertyDescriptorDisplayer.id === typeDescriptorDisplayer.id) {
+        highLevelDisplayer = $.extend({}, typeDescriptorDisplayer, propertyDescriptorDisplayer);
+      } else {
+        highLevelDisplayer = $.extend({}, propertyDescriptorDisplayer);
+      }
+      // displayer config
+      var displayerId = highLevelDisplayer.id;
+      var displayer = this.data.meta.displayers.find(function (displayer) {
+        return displayer.id === displayerId;
+      });
+      // default displayer config
+      var defaultDisplayer = { id: this.data.meta.defaultDisplayer };
+      // merge all displayers
+      var mergedDisplayer = $.extend({}, defaultDisplayer, displayer, highLevelDisplayer);
```

Comment:
The defaultDisplayer.id only has the `id` property, and is going to be overridden by the highLevelDisplayer.id if it is defined, so there shouldn't be any problems here.
The only risk I could see is someone changes this default displayer directly in the code, adding other properties than the id, but doing that is not right at all.

I coded it this way because for me the code is much more simple (only one merge), but I can update it if you're uncomfortable with that.

Reply:
Oh also there is the case the `id` is not specified in the hightLevelDisplayer, but there are some other config properties. In that case, the defaultDisplayer is going to inherit those properties, even if it does not use them. Indeed this would be cleaner to don't inherit them at all in that case!